Reading/writing goto binaries: consistently use return values

tautschnig opened the pull request:
Always use the return value to communicate errors, not a mix of return values and exceptions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Reading/writing goto binaries: consistently use return values

Always use the return value to communicate errors, not a mix of return
values and exceptions.

Terminate the "Failed to open" diagnostic with messaget::eom, consistent
with the version-error messages.

Add a unit test exercising the version-validation error branches of
write_goto_binary (unreachable via the CLI, as every production caller
uses the default version): writing with an unsupported version returns
true and -- since validation now happens before the header is written --
leaves the output stream empty, while the supported version is written
successfully.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Reading/writing goto binaries: consistently use return values

Always use the return value to communicate errors, not a mix of return
values and exceptions.

peterschrammel commented on src/goto-programs/write_goto_binary.cpp:
```diff
@@ -169,15 +171,19 @@ bool write_goto_binary(
+    message.error() << "version " << version << " no longer supported; "
+                    << "supported version = " << GOTO_BINARY_VERSION
+                    << messaget::eom;
+    return true;
```
Is this an improvement?

tautschnig replied:
Using return values rather than exceptions? I'd argue yes for two reasons:
1) It is consistent across reading (which already did this) and writing (which now does this) goto binaries.
2) Using return values to communicate errors makes it easier to build (cross-language) APIs.

tautschnig committed:
Validate goto binary version before writing header

The version validation was performed after writing the binary header to
the output stream. On an unsupported version, this would leave a
partially written (corrupt) file. Move the validation before any writes
to the stream.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Validate goto binary version before writing header

The version validation was performed after writing the binary header to
the output stream. On an unsupported version, this would leave a
partially written (corrupt) file. Move the validation before any writes
to the stream.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
